Allow transient list values to contain NULLs

avikivity opened the pull request:
The CQL protocol and specification call for lists with NULLs in
some places. For example, the statement:

```cql
UPDATE tab
SET x = 3
IF y IN (1, 2,  NULL)
WHERE pk = 4
```

has a list `(1, 2, NULL)` that contains NULL. Although the syntax is tuple-like, the value is a list;
consider the same statement as a prepared statement:

```cql
UPDATE tab
SET x = :x
IF y IN :y_values
WHERE pk = :pk
```

`:y_values` must have a list type, since the number of elements is unknown.

Currently, this is done with special paths inside LWT that bypass normal
evaluation, but if we want to unify those paths, we must allow NULLs in
lists (except in storage). This series does that.

avikivity commented:
(may fail in dtest)

nyh commented on concrete_types.hh:
```diff
@@ -19,6 +19,7 @@
 #include "utils/big_decimal.hh"
 
 struct empty_type_impl final : public abstract_type {
+    using native_type = empty_type_representation;
```
empty_type is also used in materialized views, in some cases, for unselected columns where we need to remember their liveness but not their value. See commit 30f721afabb4c0f8bf7c3c32e0ab02bcb36b77e4 - "virtual columns". This empty_type does get written to sstables. We should check that it doesn't get broken by this patch. We have tests for the virtual columns case - but it is very likely that none of them ever tried to flush to sstables and then read without cache and see that it still works....

avikivity replied:
I think the series doesn't change any of that. The path impacted is conversion to the native type (e.g. int32_t for int32_type), I think views use the serialized form throughout. I'll check it of course.

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -5469,7 +5469,7 @@ SEASTAR_TEST_CASE(test_null_and_unset_in_collections) {
         // The predicate that checks the message has to be a lambda to preserve source_location
         auto check_null_msg = [](std::source_location loc = std::source_location::current()) {
-            return exception_predicate::message_equals("null is not supported inside collections", loc);
+            return exception_predicate::message_matches(".*(null|NULL).*", loc);
```
Looks good, except it would be good to check the error has a reasonable type and wording - and not be some ugly long "internal error" .

avikivity replied:
The error messages are fine, I've seen most of them.

nyh commented on cdc/log.cc:
```diff
@@ -1084,7 +1103,7 @@ struct process_row_visitor {
         struct collection_visitor {
             bool _is_column_delete = false;
-            std::vector<managed_bytes_view> _deleted_keys;
+            std::vector<managed_bytes_view_opt> _deleted_keys;
```
I'm afraid I don't follow the details here. If this vector is supposed to hold keys, why is it "opt"? How can the key (not the value) be null?

avikivity replied:
If the data comes from a set, which shares the internal representation with a list, these could be opt (but never null).

We could de-opt them prior to adding them here. @kbr- any preference?

nyh commented on cql3/expr/expression.cc:
```diff
@@ -2277,15 +2287,19 @@ std::vector<managed_bytes_opt> get_elements(const cql3::raw_value& val, const ab
-        std::vector<managed_bytes_opt> cur_tuple = tuple_typ.split_fragmented(managed_bytes_view(element));
+    for (managed_bytes_opt& element : elements) {
+        if (!element) {
+            // Note: just skipping is okay here, for our caller
```
I don't understand this comment. It says skipping is okay, but doesn't skip...

avikivity replied:
The intent was, if we skipped instead of throwing, it would be fine as well (IIRC). I can double check.

nyh replied:
Oh, I see, you meant to write something like "just skipping **would have also been** okay"

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -5659,8 +5659,9 @@ SEASTAR_TEST_CASE(test_null_and_unset_in_collections) {
-        BOOST_REQUIRE_EXCEPTION(e.execute_prepared(where_in_list_marker, {list_with_null}).get(),
-                                exceptions::invalid_request_exception, check_null_msg());
+        auto msg = e.execute_prepared(where_in_list_marker, {list_with_null}).get();
+        assert_that(msg).is_rows().with_rows({});
```
Maybe the commit message should mention this change - that now IN allows NULL (which I guess this NULL is ignored, right?)

avikivity replied:
Yes. More properly, the NULL is compared against the column value. Since NULL does not equal NULL, there can be no matches, so the result is as if it was ignored.

nyh reviewed (commented):
Looks mostly good, but I left a few comments (I'm worried about the materialized-views use of "empty type").
In general I'm also worried that there's a lot of code here that is not being exercised by tests - although maybe your expr_tests.cc exercises most of the new code?

Also, where you change user-visible behaviors, like IN (..., null) and LWT, I think it would be good to know  what Cassandra does in this case. For this you'll probably need a cql-pytest test because the C++ tests do not let you run the same thing on Cassandra.

avikivity commented:
> Looks mostly good, but I left a few comments (I'm worried about the materialized-views use of "empty type"). In general I'm also worried that there's a lot of code here that is not being exercised by tests - although maybe your expr_tests.cc exercises most of the new code?

I hope to make this self-contained.

#12356 would exercise this code, but I don't want to rely on it for testing. What do you think is missing?


> 
> Also, where you change user-visible behaviors, like IN (..., null) and LWT, I think it would be good to know what Cassandra does in this case. For this you'll probably need a cql-pytest test because the C++ tests do not let you run the same thing on Cassandra.

Right.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3635/

nyh commented:
> > Looks mostly good, but I left a few comments (I'm worried about the materialized-views use of "empty type"). In general I'm also worried that there's a lot of code here that is not being exercised by tests - although maybe your expr_tests.cc exercises most of the new code?
> 
> I hope to make this self-contained.
> 
> #12356 would exercise this code, but I don't want to rely on it for testing. What do you think is missing?

I think I understand now - this patch, like many previous expression-related patches, adds code which doesn't actually get used yet, so it includes its own unit tests for exercising the new code outside the context of CQL. 

On one hand I agree that this is a reasonable approach. On the other hand, I think that there is a risk that some of these functions, when finally get used, may do something which makes sense on its own, but isn't 100% compatible with what Scylla did in the past or Cassandra does today, and we won't even know because our CQL test coverage in this area **might** not be great. I'm saying **might** because I really have no idea - it's possible we have great tests in this area. But it's not something you (unless I'm misunderstanding) looked into today, while you're developing the new code and everything is fresh in your mind. I would personally have preferred to write higher-level tests which can be run not just on the new expression code but also on older Scylla and on Cassandra, to make sure nothing unexpected changed during the refactoring.

Anyway, I think your approach is fine even if not the one I would have personally chosen, so I can commit this PR as soon as we understand that nothing broke regarding the empty_type use in materialized views. Tell me if you want me to write a test for this - both materialized views and test-writing is fairly fresh in my mind.

avikivity commented:
> > > Looks mostly good, but I left a few comments (I'm worried about the materialized-views use of "empty type"). In general I'm also worried that there's a lot of code here that is not being exercised by tests - although maybe your expr_tests.cc exercises most of the new code?
> > 
> > 
> > I hope to make this self-contained.
> > #12356 would exercise this code, but I don't want to rely on it for testing. What do you think is missing?
> 
> I think I understand now - this patch, like many previous expression-related patches, adds code which doesn't actually get used yet, so it includes its own unit tests for exercising the new code outside the context of CQL.

It's not that the code is unused. The code gets exercised in normal list and set processing, but the user-visible behavior changes are limited, since it mostly moves checks from closer to ingestion (when parsing a list) towards the back (just before insertion into storage).

The user visible changes are that `WHERE x IN [list with NULL]` is now accepted. In practice it makes small difference since the NULL won't match anything.


> 
> On one hand I agree that this is a reasonable approach. On the other hand, I think that there is a risk that some of these functions, when finally get used, may do something which makes sense on its own, but isn't 100% compatible with what Scylla did in the past or Cassandra does today, and we won't even know because our CQL test coverage in this area **might** not be great. I'm saying **might** because I really have no idea - it's possible we have great tests in this area. But it's not something you (unless I'm misunderstanding) looked into today, while you're developing the new code and everything is fresh in your mind. I would personally have preferred to write higher-level tests which can be run not just on the new expression code but also on older Scylla and on Cassandra, to make sure nothing unexpected changed during the refactoring.

There is no unused code. 

We do deviate from Cassandra here, Cassandra rejects NULL in lists (but accepts NULL in LWT IF clause).

> 
> Anyway, I think your approach is fine even if not the one I would have personally chosen, so I can commit this PR as soon as we understand that nothing broke regarding the empty_type use in materialized views. Tell me if you want me to write a test for this - both materialized views and test-writing is fairly fresh in my mind.

avikivity commented:
Update: 
 - clarified comment about skipping NULLs in get_list_of_tuples_elements()
 - expanded changelog when relaxing test cases
 - added serialization/deserialization test for empty type
 - added serialization/deserialization test for list type, demonstating NULL serialialization/deserialization
 - s/read_collection_value_nonnull/read_collection_value/ in one more place
 - add materialized view test with empty types to show serialization/deserialization still works

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3659/

avikivity commented:
Update:
 - added patch to relax expr::evaluate_list() too, which in turn made more test cases pass instead of throw
 - added test case (same patch) to make sure LWT rejects inserts to lists with NULL
 - fixed write_collection_value() bug uncovered by new patch
 - rebased

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3661/

nyh commented on test/cql-pytest/test_materialized_view.py:
```diff
@@ -740,3 +740,25 @@ def test_mv_override_clustering_order_error(cql, test_keyspace):
+        # Use a synchronous view so we don't have to worry about races between flush and
+        # view updates.
+        with new_materialized_view(cql, table=base, select='id', pk='id', where='id IS NOT NULL',
+                extra='WITH synchronous_updates = true') as view:
```
This test fails on Cassandra, because "synchronous_updates" is a Scylla-only invention, and even without it, Cassandra doesn't have the "virtual columns" liveness fix that we have so will likely fail on this test (I didn't check). So I think you should add the `scylla_only` fixture to this test.

nyh reviewed (changes requested):
Looks good, I just have one request about the test (which fails on Cassandra, so should be marked as such).

avikivity commented:
Update:
 - rebased past d6e6820f333aa19e82bc4c41ec8b8ee9e2ac93e2 and 100c9d068da6dffc6266ad7c3226592b871b1206
 - marked as draft because 100c9d068da6dffc6266ad7c3226592b871b1206 is still unmerged
 - marked new materialized view test as scylla_only as it cannot run on Cassandra

avikivity commented:
Update: 
 - rebased past 0b418fa7cf164e7fcbeab2d4ab1a3412eb8aeef0

avikivity commented:
success!

avikivity commented:
@nyh ping
